src/core/utils/Array.hpp

```diff
@@ -0,0 +1,78 @@
+#ifndef UTILS_ARRAY_HPP
+#define UTILS_ARRAY_HPP
+
+#include <cstddef>
+
+#if defined(__CUDACC__)
+#define DEVICE_QUALIFIER __host__ __device__
+#else
+#define DEVICE_QUALIFIER
+#endif
+
+namespace Utils {
+
+template <typename T, std::size_t N> struct Array {
+  using value_type = T;
+  using size_type = std::size_t;
+  using difference_type = std::ptrdiff_t;
+  using reference = value_type &;
+  using const_reference = const value_type &;
+  using iterator = value_type *;
+  using const_iterator = const value_type *;
+  using pointer = value_type *;
+  using const_pointer = const value_type *;
+  value_type m_data[N ? N : 1];
+  DEVICE_QUALIFIER constexpr reference at(size_type i) {
+    if (i >= N)
+      throw std::out_of_range("Array access out of bounds.");
+    return m_data[i];
+  }
+  DEVICE_QUALIFIER constexpr const_reference at(size_type i) const {
+    if (i >= N)
+      throw std::out_of_range("Array access out of bounds.");
+    return m_data[i];
+  }
+  DEVICE_QUALIFIER constexpr reference operator[](size_type i) {
+    return m_data[i];
+  }
+  DEVICE_QUALIFIER constexpr const_reference operator[](size_type i) const {
+    return m_data[i];
+  }
+  DEVICE_QUALIFIER constexpr reference front() { return *begin(); }
+  DEVICE_QUALIFIER constexpr const_reference front() const { return *cbegin(); }
+  DEVICE_QUALIFIER constexpr reference back() {
```

Comment:
Can you please check if 
```c++
std::enable_if<N > 0, reference> back() { ... }
```
works? The you could also remove the data member for `N == 0`.

Reply:
Ah, for this to work back hast to be a template...

Reply:
```warning: zero size arrays are an extension```

Reply:
Yeah, the really absolutely proper way to to it is to but into a base class and specialize that, which then also gives you empty base optimization. Like so (warninge: not tested):

```c++
template<class T, size_t N>
struct array_storage {
  T m_data[N];
};

template<class T>
struct array_storage<T, 0> {

};

template<class T, size_t N>
class Array : private array_storage<T, N> {
...
};
```

Reply:
This actually does not work if we want to preserve aggregate initialization.

Reply:
Yes, but you can make it a member.